src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -161,17 +161,19 @@ private PerAnalyzerState GetAnalyzerState(DiagnosticAnalyzer analyzer)
             }
         }
 
-        private IEnumerable<ISymbol> GetDeclaredSymbolsInTree(
+        private ImmutableArray<ISymbol> GetDeclaredSymbolsInTree(
             SyntaxTree tree,
             Compilation compilation,
             Func<SyntaxTree, Compilation, CancellationToken, SemanticModel> getCachedSemanticModel,
             CancellationToken cancellationToken)
         {
             var model = getCachedSemanticModel(tree, compilation, cancellationToken);
             var fullSpan = tree.GetRoot(cancellationToken).FullSpan;
-            var declarationInfos = new List<DeclarationInfo>();
-            model.ComputeDeclarationsInSpan(fullSpan, getSymbol: true, builder: declarationInfos, cancellationToken: cancellationToken);
-            return declarationInfos.Select(declInfo => declInfo.DeclaredSymbol).Distinct().WhereNotNull();
+            var declarationInfoBuilder = ArrayBuilder<DeclarationInfo>.GetInstance();
+            model.ComputeDeclarationsInSpan(fullSpan, getSymbol: true, builder: declarationInfoBuilder, cancellationToken: cancellationToken);
+            ImmutableArray<ISymbol> result = declarationInfoBuilder.Select(declInfo => declInfo.DeclaredSymbol).Distinct().WhereNotNull().ToImmutableArray();
```

Comment:
>WhereNotNull() [](start = 123, length = 14)

Not blocking: It might make sense to move filtering before the Distinct.

Reply:
Sure, I will address it next time this is being touched.